### PR TITLE
Chore: Turn on code coverage

### PR DESCRIPTION
**Pull request by jbrower95**
**Summary**:
This is a simple PR that enables code coverage using `forge coverage`. The only non-trivial piece is that forge coverage includes tests, scripts and other non-functional code as part of its coverage, so there's logic to strip these files out. Once the coverage site is generated, it's uploaded to S3 using an artifact action.

**How to view**:
- The bot will automatically post the coverage on your PR. You can look through the associated Github Action if you want the HTML report.

**Changelog**:
- bump hardhat version
- introduce a `coverage` job for GHA
- coverage job automatically computes coverage percentages, filtering out unnecessary files.
- coverage job then uploads report to s3 for review. 
- bot then posts the status on the PR.

### CODE REVIEW

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    - name: Install Foundry
+      uses: foundry-rs/foundry-toolchain@v1
+      with:
+        version: nightly
+    - name: Run coverage
+      run: forge coverage --report lcov
+      env:
+        RPC_MAINNET: ${{ secrets.RPC_MAINNET }}
+        RPC_HOLESKY: ${{ secrets.RPC_HOLESKY }}
+        CHAIN_ID: ${{ secrets.CHAIN_ID }}
```
Comment (ypatil12) [inline .github/workflows/coverage.yml:39]: What's chainID here?

**Review by ypatil12 — COMMENTED** [review]
Where will we see the link?

File: .github/workflows/coverage.yml (hunk 1 above)
Comment (jbrower95) [reply .github/workflows/coverage.yml]: these were empty -- we're running with default values here

Comment (jbrower95) [reply .github/workflows/coverage.yml]: updated

**Review by seanmcgary — APPROVED** [review]

**Pull request merged**